projects/js-toolkit/packages/js-api/data-set/index.ts

```diff
@@ -19,26 +21,31 @@ export interface FDSCellRenderer {
 	(args: FDSCellRendererArgs): HTMLElement;
 }
 
-export interface FDSFilterData {
-	id: string;
-	odataFilterString: string;
-	selectedData: string;
+// Frontend data set filter
+
+export interface FDSFilterData<T> {
+	selectedData: T;
+}
+
+export interface FDSFilterHTMLElementBuilderArgs<T> {
+	filter: FDSFilterData<T>;
+	setFilter: (partialFilter: Partial<FDSFilterData<T>>) => void;
 }
 
-export interface FDSFilterArgs {
-	filter: FDSFilterData;
-	setFilter: (val: Partial<FDSFilterData>) => void;
+export interface FDSFilterHTMLElementBuilder<T> {
+	(args: FDSFilterHTMLElementBuilderArgs<T>): HTMLElement;
```

Comment:
We have this [IHTMLElementBuilder interface locally in DXP](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/renderer/types.ts#L6-L8). It probably belongs here, something like:
```ts
export interface HTMLElementBuilder {
	(args: any): HTMLElement;
}

export interface FDSTableCellHTMLElementBuilder extends HTMLElementBuilder {
	(args: FDSTableCellHTMLElementBuilderArgs): HTMLElement;
}

export interface FDSFilterHTMLElementBuilder<T> extends HTMLElementBuilder {
	(args: FDSFilterHTMLElementBuilderArgs<T>): HTMLElement;
}
```

Reply:
This falls in part :two: of my summary in the big comment below, most probably.

However, I would refrain for now from creating the `HTMLElementBuilder` type, even though it is very probable that we may add it in the future.

But until I see it used in more CX types and people have started implementing and using them, I prefer not to add it since it is not absolutely necessary :shrug:

Reply:
Makes sense, at least until rule of three kicks in :+1: